pennylane/templates/subroutines/permute.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``permute`` template.
+"""
+
+import pennylane as qml
+
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def Permute(permutation, wires):
+    r"""Applies a permutation to a set of wires.
+
+    Example usage:
+
+    .. code-block:: python
+
+        import pennylane as qml
+
+        with qml.tape.QuantumTape() as tape:
+            # Send contents of wire 4 to wire 0, of wire 2 to wire 1, etc.
+            qml.templates.Permute([4, 2, 0, 1, 3], wires=[0, 1, 2, 3, 4])
+
+    See "Usage Details" below for further examples.
+
+    Args:
+        permutation (list): A list of wire labels that represents the new ordering of wires
```

Comment:
Another question, is this notation for the permutation standard?

Looking [here](https://en.wikipedia.org/wiki/Permutation#Notations), it seems like we normally specify where the wires go to, rather than where they come from. E.g., `42013` could mean `0 -> 4, 1 -> 2, 2 -> 0, 3 -> 1, 4 -> 3` rather than how it is now with `0 -> 2, 1 -> 3, 2 -> 1, 3 -> 4, 4 -> 0`.

Potentially either could be fine, but was wondering if we had a motivation to do it in the current way? And maybe we should add a note that users *shouldn't* think of it in the other way.

Reply:
Huh 🤔 

I have to admit, my first inclination is to interpret `Permute([4, 2, 0, 1, 3], wires=[0, 1, 2, 3, 4])` as moving wire 4 to the 0th position, wire 2 to the 1st position, etc.

Reply:
Good question... the rationale for doing it this way is because it looks more like a re-ordering of the wires, and that's help if the wires are not ordered in a standard way. For example, if for some reason the wires are named like ['alice', 'bob', 0, 1, 2], and you want them to be reordered as [0, 1, 2, 'alice', 'bob'], I think it's clearer to write it like that rather than [1, 2, 'alice', 'bob', 0] as would be the case with the regular notation (perm('alice') -> 1).

Reply:
I'm happy with either as long as it's clear in the docstring how it works

Reply:
Do you find it clear from how the description of the arguments are expressed? Or moreso from the examples?

Reply:
I think it was the comment in the examples that made it clear! I found the argument description a bit more ambiguous.

But I’m very much example-biased when I learn new software, so maybe it was just me?

Reply:
Thanks that seems like a great justification for the current approach! I think the examples make it clear, it's harder to see with just the docstring, but it's probably just one of those things that's clearest with an example!